src/dune_rules/glob_files.ml

```diff
@@ -0,0 +1,166 @@
+  type t =
+    { glob : Glob.t
+    ; dir : Glob_dir.t
+    ; recursive : bool
+    }
+
+  (* Returns a list of pairs (file_selector, relative_path), which correspond
+     to each directory that will be searched for files matching the glob. If the
+     glob is not recursive, this list will be of length 1. The returned file
+     selectors will expand globs relative to [base_dir], and the corresponding
+     relative paths are the paths to each directory relative to [base_dir]. The
+     relative paths are required to construct relative paths to the files found
+     by expanding the glob. *)
+  let file_selectors_with_relative_dirs { glob; dir; recursive } ~loc =
+    match (dir : Glob_dir.t) with
+    | Relative { relative_dir; base_dir } ->
+      let make_file_selector relative_dir =
+        let dir = Path.relative base_dir relative_dir in
+        File_selector.of_glob ~dir glob
+      in
+      if recursive then
+        Memo.map
+          (get_descendants_of_relative_dir_relative_to_base_dir_local ~base_dir
+             ~relative_dir)
+          ~f:
+            (List.map ~f:(fun relative_dir ->
+                 (make_file_selector relative_dir, relative_dir)))
+      else Memo.return [ (make_file_selector relative_dir, relative_dir) ]
+    | Absolute dir ->
+      if recursive then
+        User_error.raise ~loc
+          [ Pp.textf "absolute paths in recursive globs are not supported" ]
+      else
+        Memo.return
+          [ (File_selector.of_glob ~dir:(Path.of_string dir) glob, dir) ]
+end
+
+module Make_expand_vars (M : Monad.S) = struct
+  (* Functor for generating the function that expands [String_with_vars.t]s
+     into [Glob.t]s *)
+  let expand_vars { glob; recursive } ~f ~base_dir =
+    let open M.O in
+    let loc = String_with_vars.loc glob in
+    let+ glob_str = f glob in
+    let parent_str, pattern_str =
+      split_glob_string_into_parent_and_pattern glob_str
+    in
+    let glob = Glob.of_string_exn loc pattern_str in
+    let dir =
+      if Filename.is_relative parent_str then
+        Glob_dir.Relative { relative_dir = parent_str; base_dir }
+      else Glob_dir.Absolute parent_str
+    in
+    { Without_vars.glob; dir; recursive }
+end
+
+module Expand_vars_action_builder = Make_expand_vars (Action_builder)
+module Expand_vars_memo = Make_expand_vars (Memo)
+
+let expand_action_builder t ~f ~base_dir =
+  let open Action_builder.O in
+  let loc = String_with_vars.loc t.glob in
+  let* without_vars = Expand_vars_action_builder.expand_vars t ~f ~base_dir in
+  Without_vars.file_selectors_with_relative_dirs without_vars ~loc
+  |> Action_builder.of_memo
+  >>= Action_builder.List.concat_map ~f:(fun (file_selector, relative_dir) ->
+          Action_builder.paths_matching ~loc file_selector
+          >>| Path.Set.to_list
+          >>| List.map ~f:(replace_path_dir relative_dir))
+  >>| List.sort ~compare:String.compare
+
+let expand_memo t ~f ~base_dir =
```

Comment:
This function looks awfully similar, can we only keep one if possible?

Reply:
I've had a go at combining them with a functor, and added a `Glob_file.Expand` signature to avoid repetition in the interface. Can you take a look and tell me if there's too much indirection for it to be readable? It's non-trivial because it needs to generalize over the `Memo.t` and `Action_builder.t` type constructors. Is there are more ergonomic way of doing this than what I've come up with? I tried to simplify it with first-class modules but ran into an issue because I couldn't access internal parameterized types of first-class modules in function signatures.